klippa/src/serialize.rs

```diff
@@ -0,0 +1,394 @@
+    width: LinkWidth,
+    is_signed: bool,
+    whence: OffsetWhence,
+    bias: u32,
+    position: u32,
+    objidx: ObjIdx,
+}
+
+#[derive(Default)]
+#[allow(dead_code)]
+pub(crate) struct Serializer {
+    pub(crate) start: usize,
+    pub(crate) end: usize,
+    pub(crate) head: usize,
+    pub(crate) tail: usize,
+    // TODO: zerocopy, debug_depth
+    pub(crate) errors: SerializeErrorFlags,
+
+    data: Vec<u8>,
+    object_pool: ObjectPool,
+    // index for current Object in the object_pool
+    current: PoolIdx,
+
+    packed: Vec<ObjIdx>,
+    packed_map: FnvHashMap<usize, ObjIdx>,
+}
+
+#[allow(dead_code)]
+impl Serializer {
+    pub(crate) fn new(size: u32) -> Self {
+        let buf_size = size as usize;
+        let mut this = Serializer {
+            data: vec![0; buf_size],
+            end: buf_size,
+            tail: buf_size,
+            current: NULL_POOL_IDX,
+            packed: Vec::new(),
+            ..Default::default()
+        };
+
+        this.packed.push(INVALID_OBJ_IDX);
+        this
+    }
+
+    // Embed a single Scalar type
+    pub(crate) fn embed(&mut self, obj: impl Scalar) -> Result<usize, SerializeErrorFlags> {
+        let raw = obj.to_raw();
+        let bytes = raw.as_ref();
+        let size = bytes.len();
+
+        let ret = self.allocate_size(size)?;
+        self.data
+            .get_mut(ret..ret + size)
+            .unwrap()
+            .copy_from_slice(bytes);
+        Ok(ret)
+    }
+
+    // Allocate size
+    pub(crate) fn allocate_size(&mut self, size: usize) -> Result<usize, SerializeErrorFlags> {
+        if self.in_error() {
+            return Err(self.errors);
+        }
+
+        if size > u32::MAX as usize || self.tail - self.head < size {
+            return Err(self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OUT_OF_ROOM));
+        }
+
+        //TODO: add support for clear?
+        let ret = self.head;
+        self.head += size;
+        Ok(ret)
+    }
+
+    #[inline]
+    pub(crate) fn successful(&self) -> bool {
+        !self.errors
+    }
+
+    #[inline]
+    pub(crate) fn in_error(&self) -> bool {
+        !!self.errors
+    }
+
+    #[inline]
+    pub(crate) fn set_err(&mut self, error_type: SerializeErrorFlags) -> SerializeErrorFlags {
+        self.errors |= error_type;
+        self.errors
+    }
+
+    pub(crate) fn push(&mut self) -> Result<(), SerializeErrorFlags> {
+        if self.in_error() {
+            return Err(self.errors);
+        }
+
+        let pool_idx = self.object_pool.alloc();
+        let Some(obj) = self.object_pool.get_obj(pool_idx) else {
+            return Err(self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER));
+        };
+
+        obj.head = self.head;
+        obj.tail = self.tail;
+        obj.next_obj = self.current;
+        self.current = pool_idx;
+
+        Ok(())
+    }
+
+    pub(crate) fn copy_bytes(mut self) -> Vec<u8> {
+        if !self.successful() {
+            return Vec::new();
+        }
```

Comment:
I think this should be returning an error, and then the caller can choose to substitute an empty vec as needed?